### PR TITLE
[BOUNTY] Add Qwen1.5 0.5B

### PR DESCRIPTION
**Pull request by JushBJJ**
This PR is with the help of @marty1885 and @JonathanALevine.
**Bounty:** https://github.com/tenstorrent/tt-buda-demos/issues/20

The following issues need to be addressed:
- [ x ] https://github.com/tenstorrent/tt-tvm/pull/1
- [ x ] https://github.com/tenstorrent/tt-tvm/pull/2
- [ x ] https://github.com/tenstorrent/tt-tvm/issues/3
- [ x ] https://github.com/tenstorrent/tt-tvm/pull/5
- [ x ] https://github.com/tenstorrent/tt-tvm/pull/4
- [ x ] https://github.com/tenstorrent/tt-buda/issues/42
---

Video of it working: https://www.youtube.com/watch?v=Tlb1ElBHqVk

### Requirements
- Transformers >=4.42.0
- ~12 GB of RAM (Host)

### CODE REVIEW

**Comment by JushBJJ** [comment]
With all these patches, we internally have Qwen1.5 0.5B working on a Grayskull e150 (thanks @JonathanALevine). We're waiting for @marty1885 to test all of this in his e75.

@nvukobratTT @milank94

**Comment by JushBJJ** [comment]
A lot of optimizations will need to be made to speed this model up, stay tuned. I gotta do my uni assignments which are due tonight 😭

**Comment by nvukobratTT** [comment]
Great work @JushBJJ  @marty1885 @JonathanALevine! Let us know once you get silicon results out of E75 :))) 

We're working out some internal checking, as soon as we finalize those on our side we're going to complete code reviews and start with PR approvals.

Once again great progress and thanks for hunting these bounties! :)))

**Comment by marty1885** [comment]
I think there's some problems with the e75 - I ran inference for an hour but it never finished. Hope this doesn't affect the bounty as it works on the e150. It feels more like a low level compilation bug.

**Comment by JushBJJ** [comment]
@nvukobratTT @milank94 Update :)

**Comment by milank94** [comment]
@JushBJJ way to go!

As we don't have automated CI setup yet for this project, we'll manually test the model.

Before that, there are a few miscellaneous components that should be added in such as update support table, add test case, add license header, etc. You can see a breakdown here: https://github.com/tenstorrent/tt-buda-demos/blob/main/CONTRIBUTING.md#adding-models

Mind adding these details in and we'll continue the review process?

**Comment by JushBJJ** [comment]
Sure I'll do it too with the Phi 2 bounty as well, let me know if you guys had any issue running it.
Also before manually testing it, add these workarounds: 
https://github.com/JushBJJ/tt-buda/commit/f7658386d2e988c34999f3c462fe30b48fe58c36

I raised an issue (https://github.com/tenstorrent/tt-buda/issues/42) because I'm not quite sure how a proper implementation would look like for DynamicCaching so it would be great if anyone in the TT team familiar with it can make a proper fix :)

**Comment by milank94** [comment]
> Sure I'll do it too with the Phi 2 bounty as well, let me know if you guys had any issue running it. Also before manually testing it, add these workarounds: [JushBJJ/tt-buda@f765838](https://github.com/JushBJJ/tt-buda/commit/f7658386d2e988c34999f3c462fe30b48fe58c36)
> 
> I raised an issue ([tenstorrent/tt-buda#42](https://github.com/tenstorrent/tt-buda/issues/42)) because I'm not quite sure how a proper implementation would look like for DynamicCaching so it would be great if anyone in the TT team familiar with it can make a proper fix :)

Thanks for the update. We'll get someone to review and integrate the work arounds. We can look to package this as an alpha release @staylorTT 

FYI: @Shubhamsaboo for bounty tracking

**Comment by JushBJJ** [comment]
Made some changes:
- Cleaned up code and made and added batching like the other examples
- Renamed filename for qwen
- Adjusted parameters to reduce weird token generation
- Added chat version of qwen1.5

I also tested on previous transformers versions, seems like v4.42.0 is some magic number??? Any version below that Qwen does not work at all.

**Comment by JushBJJ** [comment]
Made some changes:
- Cleaned up code and added batching like the other examples
- Renamed filename for qwen
- Adjusted parameters to reduce weird token generation
- Added chat version of qwen1.5
- Added test case for qwen1.5

I also tested on previous transformers versions, seems like v4.42.0 is some magic number??? Any version below that Qwen does not work at all.

**Comment by milank94** [comment]
> Made some changes:
> 
> * Cleaned up code and added batching like the other examples
> * Renamed filename for qwen
> * Adjusted parameters to reduce weird token generation
> * Added chat version of qwen1.5
> * Added test case for qwen1.5
> 
> I also tested on previous transformers versions, seems like v4.42.0 is some magic number??? Any version below that Qwen does not work at all.

Thanks @JushBJJ . I will have a look.

That's good to note. The latest release of Buda is on `transformers==4.41.0`, so will need to uplift it in tt-buda or add it here as a temporary requirement.

File: model_demos/nlp_demos/qwen1_5/pytorch_qwen1_5.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import os
```
Comment (milank94) [inline model_demos/nlp_demos/qwen1_5/pytorch_qwen1_5.py:1]: Can you also add the following license header in this file:
```python
# SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
# SPDX-License-Identifier: Apache-2.0
```

File: model_demos/nlp_demos/qwen1_5/pytorch_qwen1_5_chat.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+import os
```
Comment (milank94) [inline model_demos/nlp_demos/qwen1_5/pytorch_qwen1_5_chat.py:1]: Can you also add the following license header in this file:
```python
# SPDX-FileCopyrightText: © 2024 Tenstorrent AI ULC
# SPDX-License-Identifier: Apache-2.0
```

**Comment by JushBJJ** [comment]
See https://github.com/tenstorrent/tt-buda/issues/42#issuecomment-2258909588

**Comment by JushBJJ** [comment]
This should be good now for final review and testing

**Comment by JushBJJ** [comment]
Any updates on this on TT's side? @milank94

**Comment by milank94** [comment]
> Any updates on this on TT's side? @milank94

Hey @JushBJJ -- we're currently working on integrating all of necessary changes into a new Buda release. This should add the support for this model and the Phi-2, along with some additional features.

Once we have that RC ready, we'll do the test and merging of these models.

Thanks for the patience.

**Comment by milank94** [comment]
@JushBJJ thank you for your patience as we're close to releasing a new release of pybuda with the changes to support this model demo. Can you switch the target for this PR to be `track_pybuda`? https://github.com/tenstorrent/tt-buda-demos/tree/track_pybuda

The process flow is that we use `track_pybuda` in our internal CI to test all of the demos and changes, then we merge those into `main`.

**Comment by milank94** [comment]
Hey @JushBJJ can you switch the target branch to be: [mkordic/rc_20240830](https://github.com/tenstorrent/tt-buda-demos/tree/mkordic/rc_20240830)?

**Comment by JushBJJ** [comment]
Done

**Comment by milank94** [comment]
Looks great @JushBJJ - can you address the merge conflict around `model_demos/pyproject.toml`?

I merged the changes for #117, I think this branch just needs a rebase onto [mkordic/rc_20240830](https://github.com/tenstorrent/tt-buda-demos/tree/mkordic/rc_20240830).

The plan is to merge these two models into a RC branch (mkordic/rc_20240830), that we are using to updated the changes for the next TT-Buda release, then merge into `main` and complete the bounties.

**Review by milank94 — APPROVED** [review]

**Comment by JushBJJ** [comment]
@milank94 Rebased it, i normally dont do it so I hope i did it right 😅

**Pull request merged**